### PR TITLE
cleanup CUDA 10 defaulted methods

### PR DESCRIPTION
**Pull request by ibaned**
The main objective of this pull request is to ensure that under CUDA 10, Kokkos does not add `__host__ __device__` to defaulted member methods, since they truly are not needed in CUDA 10. CUDA 9, however, says it doesn't need them but does need them, so we still add them in CUDA 9. This PR also includes a commit to remove the old workaround to avoid the CUDA 9 warning, since we also add a flag in our build system to silence that warning so we shouldn't need the tedious workaround too.

### CODE REVIEW

**Comment by masterleinad** [comment]
I guess you only want to keep the last two commits.

**Comment by ibaned** [comment]
Yes, I'm not sure what I should do between `develop` and the release candidate branch. Should I rebase onto develop and still target develop?

**Comment by masterleinad** [comment]
> Should I rebase onto develop and still target develop?

Yes, and then drop all the other commits.

**Comment by ibaned** [comment]
Thanks! I think its better now

**Comment by masterleinad** [comment]
You will need to fix the indentation:
```cpp
+ ./scripts/docker/check_format_cpp.sh
--- a/core/src/Kokkos_Array.hpp
+++ b/core/src/Kokkos_Array.hpp
@@ -276,7 +276,7 @@
 #ifdef KOKKOS_CUDA_9_DEFAULTED_BUG_WORKAROUND
   KOKKOS_INLINE_FUNCTION ~Array() {}
 #else
-  KOKKOS_DEFAULTED_FUNCTION ~Array()                       = default;
+  KOKKOS_DEFAULTED_FUNCTION ~Array() = default;
 #endif
   Array()                 = delete;
   Array(const Array& rhs) = delete;
@@ -349,7 +349,7 @@
 #ifdef KOKKOS_CUDA_9_DEFAULTED_BUG_WORKAROUND
   KOKKOS_INLINE_FUNCTION ~Array() {}
 #else
-  KOKKOS_DEFAULTED_FUNCTION ~Array()                       = default;
+  KOKKOS_DEFAULTED_FUNCTION ~Array() = default;
 #endif
   Array()             = delete;
   Array(const Array&) = delete;
1 file(s) not formatted properly:
    core/src/Kokkos_Array.hpp
script returned exit code 1
```

File: core/src/Kokkos_Array.hpp [hunk 1]
```diff
@@ -283,7 +276,7 @@ struct Array<T, KOKKOS_INVALID_INDEX, Array<>::contiguous> {
 #ifdef KOKKOS_CUDA_9_DEFAULTED_BUG_WORKAROUND
```
Comment (dalg24) [inline core/src/Kokkos_Array.hpp]: Drop the `#ifdef KOKKOS_CUDA_9_DEFAULTED_BUG_WORKAROUND`

File: core/src/Kokkos_Array.hpp [hunk 2]
```diff
@@ -356,7 +349,7 @@ struct Array<T, KOKKOS_INVALID_INDEX, Array<>::strided> {
...
```
Comment (dalg24) [inline core/src/Kokkos_Array.hpp]: `#ifdef KOKKOS_CUDA_9_DEFAULTED_BUG_WORKAROUND` not needed

File: core/src/Kokkos_Pair.hpp [hunk 3]
```diff
@@ -438,7 +431,7 @@ struct pair<T1, void> {
   first_type first;
   enum { second = 0 };
 
-  KOKKOS_FORCEINLINE_FUNCTION constexpr
+  KOKKOS_DEFAULTED_FUNCTION constexpr
 #ifdef KOKKOS_CUDA_9_DEFAULTED_BUG_WORKAROUND
```
Comment (dalg24) [inline core/src/Kokkos_Pair.hpp]: Here as well

**Comment by ibaned** [comment]
Thanks @masterleinad and @dalg24 !
I just pushed some possible fixes to these comments.

**Comment by dalg24** [comment]
clang-format [failed](https://cloud.cees.ornl.gov/jenkins-ci/blue/organizations/jenkins/Kokkos/detail/Kokkos/847/pipeline/18)

**Comment by ibaned** [comment]
Thanks @masterleinad ! I'll work on adding those to the pull request

File: cmake/kokkos_arch.cmake [hunk 4]
```diff
@@ -175,7 +175,7 @@ IF (KOKKOS_CXX_COMPILER_ID STREQUAL NVIDIA)
     GLOBAL_APPEND(KOKKOS_CUDA_OPTIONS -lineinfo)
   ENDIF()
   UNSET(_UPPERCASE_CMAKE_BUILD_TYPE)
-  IF (KOKKOS_CXX_COMPILER_VERSION VERSION_GREATER 9.0 OR KOKKOS_CXX_COMPILER_VERSION VERSION_EQUAL 9.0)
+  IF (KOKKOS_CXX_COMPILER_VERSION VERSION_EQUAL 9.0)
```
Comment (masterleinad) [inline cmake/kokkos_arch.cmake]: What about 9.1 and 9.2?

Comment (ibaned) [reply cmake/kokkos_arch.cmake]: Ah shoot this will miss those

**Comment by ibaned** [comment]
Not sure what to make of this...
```
[ RUN      ] serial.view_layoutstride_left_to_layoutright_assignment
1393view_layoutstride_left_to_layoutright_assignment: srand(1578548612)
1394/home/travis/build/kokkos/kokkos/core/unit_test/TestViewLayoutStrideAssignment.hpp:829: Failure
1395Death test: { dst = src; }
1396    Result: died but not with expected error.
1397  Expected: View assignment must have compatible layouts
1398Actual msg:
1399[  DEATH   ] 
1400[  FAILED  ] serial.view_layoutstride_left_to_layoutright_assignment (118 ms)
```

**Comment by masterleinad** [comment]
This is a known problem, see #2617. I guess we can just ignore it here.

**Comment by masterleinad** [comment]
Would you mind squashing your commits a bit?

**Comment by ibaned** [comment]
Is all one commit okay?

**Comment by masterleinad** [comment]
Sure. I would have suggested either one commit or two for the changes to ` cmake/kokkos_arch.cmake` and the rest.

**Pull request merged**